# ****Midterm of Tomáš Šamaj (36902772)****

In [4]:
%matplotlib inline
import pandas as pd
import seaborn as sns
import math
import matplotlib.pyplot as plt
import scipy.stats as stats
import numpy as np
import seaborn as sns
import datetime
import zipfile

access the data

In [108]:
ADP = pd.read_csv(".\ADP.csv")
EIX = pd.read_csv(".\EIX.csv")
FDS = pd.read_csv(".\FDS.csv")
IPG = pd.read_csv(".\IPG.csv")
ticker_list = [ADP,EIX,FDS,IPG]

#set index to datetime
for ticker in ticker_list:
 ticker.set_index(["Date"],inplace = True)

ADP

High         Low        Open       Close     Volume  \
Date                                                                    
2017-11-13  112.059998  110.550003  110.980003  111.529999  3716200.0   
2017-11-14  112.230003  110.940002  111.309998  112.080002  1992000.0   
2017-11-15  112.230003  110.379997  112.199997  110.660004  1555200.0   
2017-11-16  111.480003  110.470001  111.000000  111.029999  1870200.0   
2017-11-17  111.220001  110.269997  110.980003  111.059998  1969500.0   
...                ...         ...         ...         ...        ...   
2022-11-07  241.229996  236.320007  238.520004  240.410004  1250400.0   
2022-11-08  248.449997  240.119995  240.940002  245.039993  1724200.0   
2022-11-09  245.600006  240.009995  243.649994  240.380005  1242900.0   
2022-11-10  255.669998  247.580002  251.179993  254.880005  2672800.0   
2022-11-11  255.960007  247.429993  254.210007  249.229996  2126800.0   

             Adj Close  
Date                    
2017-11-13  100.551643  
2017-11-14  101.047508  
2017-11-15   99.767288  
2017-11-16  100.100868  
2017-11-17  100.127922  
...                ...  
2022-11-07  240.410004  
2022-11-08  245.039993  
2022-11-09  240.380005  
2022-11-10  254.880005  
2022-11-11  249.229996  

[1259 rows x 6 columns]

1) Is there a company that has no difference between the Close and Adj Close columns?
What does it mean from the financial point of view for the stock (you can get bonus
partial points)?

In [44]:
print(sum(ADP["Close"]==ADP["Adj Close"])==0)
print(sum(EIX["Close"]==EIX["Adj Close"])==0)
print(sum(FDS["Close"]==FDS["Adj Close"])==0)
print(sum(IPG["Close"]==IPG["Adj Close"])==0)


False
False
False
False


Since it is not true for any ticker, there is a difference between Close and Adj Close columns for all companies.

From the financial point of view this means that during the sample period, each of the 4 companies has at least once paid dividends/split stocks or issued new shares after trading hours. In such cases, the Adj. Close price is different from the Close price of the day.

2) What is the highest and lowest price (Adj Close) each company recorded?

In [54]:
for ticker in ticker_list:
    print(ticker["Adj Close"].max(),ticker["Adj Close"].min())

259.8173217773437 98.1278076171875
72.75 40.61245727539063
487.8295593261719 178.72164916992188
39.13483428955078 11.541755676269531


ADP: max=259.8173217773437, min=98.1278076171875
EIX: max=72.75, min=40.61245727539063
FDS: max=487.8295593261719, min=178.72164916992188
IPG: max=39.13483428955078, min=11.541755676269531

3) Calculate logarithmic returns from Adj Close. For each company report on
its, min, max, mean, median of the return distribution.

In [83]:
ADP_returns = np.log(ADP["Adj Close"]/ADP["Adj Close"].shift(1))
EIX_returns = np.log(EIX["Adj Close"]/EIX["Adj Close"].shift(1))
FDS_returns = np.log(FDS["Adj Close"]/FDS["Adj Close"].shift(1))
IPG_returns = np.log(IPG["Adj Close"]/IPG["Adj Close"].shift(1))

print("ADP")
print(ADP_returns.describe())
print("EIX")
print(EIX_returns.describe())
print("FDS")
print(FDS_returns.describe())
print("IPG")
print(IPG_returns.describe())

ADP
count    1258.000000
mean        0.000722
std         0.017758
min        -0.165495
25%        -0.006691
50%         0.001459
75%         0.009388
max         0.111573
Name: Adj Close, dtype: float64
EIX
count    1258.000000
mean       -0.000102
std         0.020523
min        -0.143101
25%        -0.008925
50%         0.000598
75%         0.009729
max         0.143101
Name: Adj Close, dtype: float64
FDS
count    1258.000000
mean        0.000645
std         0.018399
min        -0.117682
25%        -0.007156
50%         0.001863
75%         0.009211
max         0.140840
Name: Adj Close, dtype: float64
IPG
count    1258.000000
mean        0.000585
std         0.022848
min        -0.165692
25%        -0.010357
50%         0.000953
75%         0.011512
max         0.136922
Name: Adj Close, dtype: float64


4) When did each company record the highest gain and highest loss for the day?
(logarithmic loss). Hint: idxmax

In [86]:
print("HIGHEST GAINS:")
print("ADP")
print(ADP_returns.idxmax())
print("EIX")
print(EIX_returns.idxmax())
print("FDS")
print(FDS_returns.idxmax())
print("IPG")
print(IPG_returns.idxmax())

HIGHEST GAINS:
ADP
2020-03-26
EIX
2018-11-16
FDS
2020-06-25
IPG
2020-03-24


In [87]:
print("HIGHEST LOSSES:")
print("ADP")
print(ADP_returns.idxmin())
print("EIX")
print(EIX_returns.idxmin())
print("FDS")
print(FDS_returns.idxmin())
print("IPG")
print(IPG_returns.idxmin())

HIGHEST LOSSES:
ADP
2020-03-16
EIX
2020-03-16
FDS
2020-03-18
IPG
2020-03-12


5) What is the average calendar weekly volume for each company? Hint: check how to
resample pandas DF

In [105]:
ADP.index = pd.to_datetime(ADP.index)
ADP_res = ADP.resample("W").sum() #Sum the daily volumes to get weekly
print("ADP")
print(ADP_res["Volume"].mean())

EIX.index = pd.to_datetime(EIX.index)
EIX_res = EIX.resample("W").sum() #Sum the daily volumes to get weekly
print("EIX")
print(EIX_res["Volume"].mean())

FDS.index = pd.to_datetime(FDS.index)
FDS_res = FDS.resample("W").sum() #Sum the daily volumes to get weekly
print("FDS")
print(FDS_res["Volume"].mean())

IPG.index = pd.to_datetime(IPG.index)
IPG_res = IPG.resample("W").sum() #Sum the daily volumes to get weekly
print("IPG")
print(IPG_res["Volume"].mean())


ADP
9543471.647509579
EIX
11305770.881226053
FDS
1395741.3793103448
IPG
19628102.681992337


Here we can see the average weekly volume for each company

6) Which company recorded the highest total return over the whole period?

In [114]:
ADP_tot = np.log(ADP["Adj Close"][-1]/ADP["Adj Close"][0])
print("ADP")
print(ADP_tot)

EIX_tot = np.log(EIX["Adj Close"][-1]/EIX["Adj Close"][0])
print("EIX")
print(EIX_tot)

FDS_tot = np.log(FDS["Adj Close"][-1]/FDS["Adj Close"][0])
print("FDS")
print(FDS_tot)

IPG_tot = np.log(IPG["Adj Close"][-1]/IPG["Adj Close"][0])
print("IPG")
print(IPG_tot)

print(max(ADP_tot,EIX_tot,FDS_tot,IPG_tot))

ADP
0.9077046878477927
EIX
-0.12888781272445232
FDS
0.8113694502214714
IPG
0.7355804325398948
0.9077046878477927
